tests/test_dependency_manager.py

```diff
@@ -142,7 +149,22 @@ def test_exclude_complex(dist_graph):
 
     pex_info = PexInfo.default()
     pex_builder = PEXBuilder(pex_info=pex_info)
-    dependency_manager.configure(pex_builder, excluded=["c"])
+    with pytest.raises(AssertionError) as exec_info:
+        dependency_manager.configure(
```

Comment:
Sorry, but I'm a bit slow on the uptake from this code. Why does this configure call fail but the later one does not?

Reply:
Because the input requirements are `dist_graph.root_reqs` (see: [147](https://github.com/pex-tool/pex/pull/2409/files/2967b7d9e133dfa7dfcaf4d65d89d69df2eb0994..5591f758ccd3def5e729d9c3ced965ed17f05b1e#diff-1734962c259079fe44ab215607c037d799acd31b8f29f8c0d0c41e880d4f4145R147)) which is [`["a", "b"]`](https://github.com/pex-tool/pex/pull/2409/files/2967b7d9e133dfa7dfcaf4d65d89d69df2eb0994..5591f758ccd3def5e729d9c3ced965ed17f05b1e#diff-1734962c259079fe44ab215607c037d799acd31b8f29f8c0d0c41e880d4f4145R91). The `--exclude`, however is for "c", which is a transitive dependency, but "c" has not been excluded from The `DependencyManager.distributions`. The `DependencyManager` only handles excluding root reqs in the case you both ask for a thing _and_ `--exclude` the thing. The tricky case being `pex /this/local/project --exclude "foo<1" ...` where Pex doesn't know the project name and version of the `/this/local/project` requirement until _after_ the Pip resolve and wheel build, which is when the `DependencyManager` finishes things up building the PEX.

The latter call doesn't fail because the `DependencyManager.distributions` has a hole - "C" is excluded already (by Pip).

Reply:
> The DependencyManager only handles excluding root reqs in the case you both ask for a thing and --exclude the thing

Ah, okay 👍 makes sense thanks.